File: klinara-ios/klinara-ios/Services/Formatting/Slug.swift
import Foundation

/// Sunucunun `SLUG_PATTERN`'ine (`^[a-z0-9]([a-z0-9-]{1,48}[a-z0-9])$`) uyan
/// kod üretimi.
///
/// Kullanıcı slug yazmak zorunda kalmasın diye addan türetilir. Türkçe
/// karakterler `Foundation`'ın genel çevirisine bırakılmaz: `ı` harfi bazı
/// dönüşümlerde tamamen düşer ve "Işıl" → "sl" gibi bir kod üretir.
enum Slug {

    private static let turkishMap: [Character: String] = [
        "ç": "c", "ğ": "g", "ı": "i", "ö": "o", "ş": "s", "ü": "u",
        "Ç": "c", "Ğ": "g", "İ": "i", "I": "i", "Ö": "o", "Ş": "s", "Ü": "u",
    ]

    static func make(from name: String) -> String {
        var output = ""
        var lastWasSeparator = true

        for character in name {
            if let mapped = turkishMap[character] {
                output += mapped
                lastWasSeparator = false
            } else if character.isLetter || character.isNumber {
                let lowered = character.lowercased()
                // Türkçe dışı aksanlar (é, ü, â) ASCII'ye indirilir.
                let ascii = lowered.folding(options: .diacriticInsensitive, locale: Locale(identifier: "en_US_POSIX"))
                let filtered = ascii.filter { $0.isASCII && ($0.isLetter || $0.isNumber) }
                if filtered.isEmpty {
                    if !lastWasSeparator { output += "-"; lastWasSeparator = true }
                } else {
                    output += filtered
                    lastWasSeparator = false
                }
            } else if !lastWasSeparator {
                output += "-"
                lastWasSeparator = true
            }
        }

        // Baştaki/sondaki tireler desene takılır.
        var trimmed = output
        while trimmed.hasPrefix("-") { trimmed.removeFirst() }
        while trimmed.hasSuffix("-") { trimmed.removeLast() }
        return String(trimmed.prefix(50))
    }

    /// Sunucu en az 3, en çok 50 karakter ister.
    static func isValid(_ slug: String) -> Bool {
        let pattern = /^[a-z0-9][a-z0-9-]{1,48}[a-z0-9]$/
        return slug.count >= 3 && slug.count <= 50 && slug.wholeMatch(of: pattern) != nil
    }
}

File: klinara-android/app/src/main/java/com/klinara/android/services/formatting/Slug.kt
package com.klinara.android.services.formatting

import java.text.Normalizer
import java.util.Locale

/**
 * Sunucunun `SLUG_PATTERN`'ine (`^[a-z0-9][a-z0-9-]{1,48}[a-z0-9]$`) uyan kod üretimi.
 *
 * Kullanıcı slug yazmak zorunda kalmasın diye addan türetilir. Türkçe karakterler genel
 * bir çeviriye BIRAKILMAZ: `ı` harfi bazı dönüşümlerde tamamen düşer ve "Işıl" → "sl"
 * gibi bir kod üretir.
 */
object Slug {
    private const val MIN_LENGTH = 3
    private const val MAX_LENGTH = 50

    private val turkishMap: Map<Char, String> =
        mapOf(
            'ç' to "c", 'ğ' to "g", 'ı' to "i", 'ö' to "o", 'ş' to "s", 'ü' to "u",
            'Ç' to "c", 'Ğ' to "g", 'İ' to "i", 'I' to "i", 'Ö' to "o", 'Ş' to "s", 'Ü' to "u",
        )

    fun make(name: String): String {
        val output = StringBuilder()
        var lastWasSeparator = true

        for (character in name) {
            lastWasSeparator = append(character, output, lastWasSeparator)
        }

        // Baştaki/sondaki tireler desene takılır.
        return output.toString().trim('-').take(MAX_LENGTH)
    }

    /**
     * Tek karakteri işler ve "son yazılan bir ayırıcı mıydı" durumunu döndürür.
     *
     * Ayrı bir fonksiyon: gövde bir karakter durum makinesi ve döngünün içinde
     * yazıldığında dört seviye iç içe bloğa çıkıyor.
     */
    private fun append(
        character: Char,
        output: StringBuilder,
        lastWasSeparator: Boolean,
    ): Boolean {
        val mapped = turkishMap[character]
        val text =
            when {
                mapped != null -> mapped
                // Latin'e indirilemeyen bir harf (örn. Kiril) ayırıcı gibi davranır.
                character.isLetterOrDigit() -> asciiFold(character)
                else -> ""
            }

        if (text.isNotEmpty()) {
            output.append(text)
            return false
        }

        // Ayırıcı: art arda tire yazılmaz.
        if (!lastWasSeparator) output.append('-')
        return true
    }

    /** Sunucu en az 3, en çok 50 karakter ister. */
    fun isValid(slug: String): Boolean =
        slug.length in MIN_LENGTH..MAX_LENGTH && PATTERN.matches(slug)

    /** Türkçe dışı aksanlar (é, â) ASCII'ye indirilir; indirilemeyen karakter düşer. */
    private fun asciiFold(character: Char): String =
        Normalizer
            .normalize(character.lowercase(Locale.ROOT), Normalizer.Form.NFD)
            .filter { it.code < ASCII_LIMIT && it.isLetterOrDigit() }

    private const val ASCII_LIMIT = 128
    private val PATTERN = "^[a-z0-9][a-z0-9-]{1,48}[a-z0-9]$".toRegex()
}
